File: src/main/java/io/github/jsnimda/common/gui/AnchoredListWidget.java
package io.github.jsnimda.common.gui;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.AbstractParentElement;
import net.minecraft.client.gui.Drawable;
import net.minecraft.client.gui.Element;
import net.minecraft.client.util.math.MatrixStack;

public class AnchoredListWidget<E extends AnchoredListWidget.Entry> extends AbstractParentElement implements Drawable {

  private List<Element> children = new ArrayList<>();
  private List<E> entries = new ArrayList<>();
  private int entriesTotalHeight = 0;
  private int x;
  private int y;
  private int width;
  private int height;
  public boolean renderBorder = true;
  public boolean showAnchorsHeader = true;
  protected int anchorHeaderHeight = 13;

  private ScrollableContainerWidget container = new ScrollableContainerWidget();

  protected AnchorHeader anchorHeader = new AnchorHeader();

  private static final int COLOR_ANCHOR_BORDER      = 0x80999999;
  private static final int COLOR_ANCHOR_BG          = 0x40999999;
  private static final int COLOR_ANCHOR_BORDER_HOVER      = 0xFF999999;
  private static final int COLOR_ANCHOR_BG_HOVER          = 0xC0999999;
  protected class AnchorHeader extends AbstractParentElement implements Drawable {
    protected class Anchor {
      public String anchorDisplayText;
      public TextButtonWidget textButtonWidget;
      public int toScrollY;
      public int rowIndex;
      public Anchor(String displayText, int toScrollY, int rowIndex) {
        this.anchorDisplayText = displayText;
        this.toScrollY = toScrollY;
        this.rowIndex = rowIndex;
        textButtonWidget = new TextButtonWidget(0, 0, displayText, x -> {
          container.setScrollY(toScrollY);
        });
        textButtonWidget.setHoverText("§e§n" + displayText); // underline
        textButtonWidget.setInactiveText("§e§n§l" + displayText); // underline + bold
        textButtonWidget.visible = false;
        textButtonWidget.pressableMargin = 2;
      }
    }
    private List<Anchor> anchors = new ArrayList<>();
    private boolean expanded = false;
    private final int SEPARATOR_WIDTH = 10;
    private final int DOT_DOT_DOT_WIDTH = MinecraftClient.getInstance().textRenderer.getWidth(" ... ...");
    private int rowHeight = 13;

    private int currentTextX = 0;
    private int currentTextRow = 0;

    private int hoveringY = -1; // reset to -1 if unhover

    private int getTextOffsetY() {
      return (rowHeight + 1) / 2 - 4;
    }

    private int getTotalTextRow() {
      return currentTextRow + 1;
    }

    private int getViewingTextRow() {
      if (anchors.isEmpty()) {
        return 0;
      }
      int index = getViewingAnchorIndex();
      return anchors.get(index).rowIndex;
    }

    private int getExpandedHeight() {
      return rowHeight * getTotalTextRow() + 1;
    }

    private int getExpandedY() {
      if (hoveringY == -1) {
        hoveringY = Math.max(7, getHiddenStartY());
      }
      return hoveringY;
    }

    private int getHiddenStartY() {
      return y - getViewingTextRow() * rowHeight;
    }

    private int getAvailableStringWidth() {
      return width - 20 - DOT_DOT_DOT_WIDTH;
    }

    private int getStartX() {
      return x + 10;
    }

    public void addAnchor(String displayText) {
      addAnchor(displayText, entriesTotalHeight);
    }
    public void addAnchor(String displayText, int toScrollY) {
      int w = MinecraftClient.getInstance().textRenderer.getWidth("§n" + displayText);
      if (w + currentTextX > getAvailableStringWidth()) {
        currentTextX = 0;
        currentTextRow++;
      }
      currentTextX += w + SEPARATOR_WIDTH;
      anchors.add(new Anchor(displayText, toScrollY, currentTextRow));
    }

    private int getViewingAnchorIndex() {
      int scrollY = container.getScrollY();
      for (int i = 0; i < anchors.size(); i++) {
        if (anchors.get(i).toScrollY >= scrollY) {
          if (anchors.get(i).toScrollY - scrollY <= container.getViewportHeight() / 2) {
            return i;
          } else {
            return Math.max(0, i - 1);
          }
        }
      }
      return anchors.size() - 1;
    }

    @Override
    public List<? extends Element> children() {
      return anchors.stream().map(x -> x.textButtonWidget).collect(Collectors.toList());
    }

    @Override
    public void render(MatrixStack matrices, int mouseX, int mouseY, float partialTicks) {
      isMouseOver(mouseX, mouseY); // update expanded state
      if (expanded) {
        int y1 = getExpandedY();
        int y2 = getExpandedY() + getExpandedHeight();
        VHLine.h(matrices, x, x + width - 1, y1, COLOR_ANCHOR_BORDER_HOVER);
        VHLine.h(matrices, x, x + width - 1, y2, COLOR_ANCHOR_BORDER_HOVER);
        fill(matrices, x, y1 + 1, x + width, y2, COLOR_ANCHOR_BG_HOVER);
      } else {
        VHLine.h(matrices, x, x + width - 1, y, COLOR_ANCHOR_BORDER);
        VHLine.h(matrices, x, x + width - 1, y + anchorHeaderHeight, COLOR_ANCHOR_BORDER);
        fill(matrices, x, y + 1, x + width, y + anchorHeaderHeight, COLOR_ANCHOR_BG);
      }
      if (!anchors.isEmpty()) {
        int startY = expanded ? getExpandedY() : getHiddenStartY();
        int viewingAnchorIndex = getViewingAnchorIndex();
        int viewingRowIndex = anchors.get(viewingAnchorIndex).rowIndex;
        int offsetX = 0;
        int lastRowIndex = 0;
        for (int i = 0; i < anchors.size(); i++) {
          Anchor anchor = anchors.get(i);
          if (anchor.rowIndex != lastRowIndex) {
            offsetX = 0;
            lastRowIndex = anchor.rowIndex;
          }
          TextButtonWidget textWidget =  anchor.textButtonWidget;
          textWidget.active = i != viewingAnchorIndex;
          textWidget.visible = expanded || anchor.rowIndex == viewingRowIndex;
          textWidget.x = getStartX() + offsetX;
          textWidget.y = startY + getTextOffsetY() + anchor.rowIndex * rowHeight;
          textWidget.render(matrices, mouseX, mouseY, partialTicks);
          offsetX += textWidget.getWidth() + SEPARATOR_WIDTH;
          if (!expanded && getTotalTextRow() > 1 && anchor.rowIndex == viewingRowIndex
              && (i + 1 >= anchors.size() || anchor.rowIndex != anchors.get(i + 1).rowIndex)) {
            // this method doesn't exist anymore, thanks 1.16
            //drawString(MinecraftClient.getInstance().textRenderer, " ... ...", getStartX() + offsetX - SEPARATOR_WIDTH, y + getTextOffsetY(), COLOR_WHITE);
            MinecraftClient.getInstance().textRenderer.draw(matrices, " ... ...", getStartX() + offsetX - SEPARATOR_WIDTH, y + getTextOffsetY(), COLOR_WHITE);
          }
        }
      }
    }

    @Override
    public boolean isMouseOver(double d, double e) {
      boolean result;
      if (!expanded) {
        result = VHLine.contains(x, y, x + width, y + anchorHeaderHeight + 1, (int)d, (int)e);
      } else {
        int y1 = getExpandedY();
        result = VHLine.contains(x, y1, x + width, y1 + getExpandedHeight(), (int)d, (int)e);
      }
      expanded = result && getTotalTextRow() > 1;
      if (!expanded) {
        hoveringY = -1;
      }
      return result;
    }

    public void sizeChanged() {
      // recalc
      List<Anchor> a = anchors;
      anchors = new ArrayList<>();
      currentTextX = 0;
      currentTextRow = 0;
      a.forEach(x -> addAnchor(x.anchorDisplayText, x.toScrollY));
    }
  }

  public AnchoredListWidget(int x, int y, int width, int height) {
    children.add(anchorHeader);
    children.add(container);
    setBounds(x, y, width, height);
    container.setContentRenderer((matrices, mx, my, p, vx, vy, vw, vh, sy) -> {
      int dy = vy - sy;
      for (E e : entries) {
        int eh = e.getHeight();
        if (dy + eh > vy && dy < vy + vh) {
          e.render(matrices, mx, my, p, vx, dy, vw);
        }
        dy += eh;
      }
    });
    container.renderBorder = false;
    renderBorder = false;
  }

  public ScrollableContainerWidget getContainer() {
    return container;
  }

  public void setBounds(int x, int y, int width, int height) {
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;
    container.setBounds(x, y + anchorHeaderHeight, width, height - anchorHeaderHeight);
    anchorHeader.sizeChanged();
  }

  @Override
  public List<Element> children() {
    return children;
  }

  private static final int COLOR_WHITE              = 0xFFFFFFFF;
  private static final int COLOR_BORDER             = 0xFF999999;
  public int borderColor = COLOR_BORDER;
  @Override
  public void render(MatrixStack matrices, int mouseX, int mouseY, float partialTicks) {
    if (renderBorder) {
      VHLine.outline(matrices, x, y, x + width, y + height, borderColor);
    }

    container.render(matrices, mouseX, mouseY, partialTicks);

    // render header
    anchorHeader.render(matrices, mouseX, mouseY, partialTicks);
  }


  public void addEntry(E entry) {
    entries.add(entry);
    children.add(entry);
    entriesTotalHeight += entry.getHeight();
    container.setContentHeight(entriesTotalHeight);
  }

  public void addAnchor(String displayText) {
    anchorHeader.addAnchor(displayText);
  }

  @Override
  public boolean isMouseOver(double d, double e) {
    return VHLine.contains(x, y, x + width, y + height, (int)d, (int)e) || anchorHeader.isMouseOver(d, e);
  }

  @Override
  public boolean mouseReleased(double d, double e, int i) {
    if (super.mouseReleased(d, e, i)) {
      return true;
    }
    return container.mouseReleased(d, e, i);
  }

  @Override
  public boolean mouseClicked(double d, double e, int i) {
    if (!isMouseOver(d, e)) {
      return false;
    }
    if (anchorHeader.isMouseOver(d, e)) {
      anchorHeader.mouseClicked(d, e, i);
      return true;
    }
    if (container.isMouseOverViewport(d, e)) {
      return super.mouseClicked(d, e, i);
    }
    if (container.isMouseOver(d, e)) {
      if (container.mouseClicked(d, e, i)) {
        this.setFocused(container); // super.mouseClicked
        if (i == 0) {
           this.setDragging(true);
        }
        return true;
      }
    }
    return false;
  }

  public static interface Entry extends Element {
    int getHeight();
    void render(MatrixStack matrices, int mouseX, int mouseY, float partialTicks, int offsetX, int offsetY, int viewportWidth);
  }


}

File: src/main/java/io/github/jsnimda/common/gui/widget/AnchoredListWidget.kt
package io.github.jsnimda.common.gui.widget

import io.github.jsnimda.common.gui.widget.FlowLayout.FlowDirection.TOP_DOWN
import io.github.jsnimda.common.gui.widget.Overflow.VISIBLE
import io.github.jsnimda.common.vanilla.VHLine.fill
import io.github.jsnimda.common.vanilla.VHLine.h
import io.github.jsnimda.common.vanilla.VHLine.outline
import io.github.jsnimda.common.vanilla.VanillaRender

private const val COLOR_ANCHOR_BORDER = -0x7f666667
private const val COLOR_ANCHOR_BG = 0x40999999
private const val COLOR_ANCHOR_BORDER_HOVER = -0x666667
private const val COLOR_ANCHOR_BG_HOVER = -0x3f666667
private const val COLOR_WHITE = -0x1
private const val COLOR_BORDER = -0x666667

private const val rowHeight = 13
private const val leastY = 7
private const val SEPARATOR_WIDTH = 10

open class AnchoredListWidget : Widget() {
  var anchorHeader = AnchorHeader()
  val container = ScrollableContainerWidget().apply {
    anchor = AnchorStyles.all
    this@AnchoredListWidget.widgets.add(this)
    top = anchorHeader.height
    right = 0
    bottom = 0
  }
  val containerContentFlowLayout = FlowLayout(container.contentContainer, TOP_DOWN)

  var renderBorder = true
  var borderColor = COLOR_BORDER

  init {
    overflow = VISIBLE
//    container.renderBorder = false
//    renderBorder = false
  }

  // ============
  // AnchorHeader
  // ============

  inner class AnchorHeader : Widget() {
    init {
      anchor = AnchorStyles.noBottom
      this@AnchoredListWidget.widgets.add(this)
      height = rowHeight
      right = 0
      zIndex = 1
    }

    fun addAnchor(displayText: String, toScrollY: Int = container.contentHeight) {
      anchorsManager.addAnchor(displayText, toScrollY)
    }

    private val dummyAnchor = Anchor("", 0, 0)

    inner class Anchor(var anchorDisplayText: String, val toScrollY: Int, var rowIndex: Int) {
      val textButtonWidget = TextButtonWidget(anchorDisplayText, { -> container.scrollY = toScrollY }).apply {
        hoverText = "§e§n$anchorDisplayText"
        inactiveText = "§e§n§l$anchorDisplayText"
        visible = false
        pressableMargin = 2
      }
    }

    val anchors: List<Anchor>
      get() = anchorsManager.anchors
    private var anchorsManager = AnchorsManager()

    private inner class AnchorsManager { // this class manage anchor adding
      private val ellipsisTextWidth = VanillaRender.getStringWidth(" ... ...")

      val width = this@AnchorHeader.width
      private val availableWidth
        get() = width - 20 - ellipsisTextWidth

      val anchors = mutableListOf<Anchor>()

      private var textLeft = 0
      private var textRowIndex = 0

      val textYPerRow
        get() = (rowHeight + 1) / 2 - 4
      val totalTextRow
        get() = textRowIndex + 1

      fun addAnchor(displayText: String, toScrollY: Int) {
        val textWidth = VanillaRender.getStringWidth("§n$displayText")
        if (textLeft + textWidth > availableWidth) {
          textLeft = 0
          textRowIndex++
        }
        textLeft += textWidth + SEPARATOR_WIDTH
        Anchor(displayText, toScrollY, textRowIndex).apply {
          anchors.add(this)
          this@AnchorHeader.widgets.add(this.textButtonWidget)
        }
      }
    }

    private var _expanded = false
    var expanded
      get() = _expanded
      set(value) {
        if (value) expand()
        else collaspe()
      }

    fun expand() {
      if (_expanded) return
      _expanded = true
      val leastTop = leastY - this@AnchoredListWidget.screenY
      this.top = (-rowHeight * highlightingAnchor.rowIndex).coerceAtLeast(leastTop)
      this.height = rowHeight * anchorsManager.totalTextRow + 1
      updateTexts()
    }

    fun collaspe() {
      if (!_expanded) return
      _expanded = false
      this.top = 0
      this.height = rowHeight
      updateTexts()
    }

    fun updateTexts() {
      val startLeft = 10
      var textLeft = startLeft
      var textTop = anchorsManager.textYPerRow
      var lastRowIndex = 0
      anchors.forEachIndexed { index, anchor ->
        anchor.textButtonWidget.visible = expanded || highlightingAnchor.rowIndex == anchor.rowIndex
        if (anchor.textButtonWidget.visible) anchor.textButtonWidget.apply {
          if (lastRowIndex != anchor.rowIndex) {
            lastRowIndex = anchor.rowIndex
            textLeft = startLeft
            textTop += if (expanded) rowHeight else 0
          }
          top = textTop
          left = textLeft
          textLeft += width + SEPARATOR_WIDTH
        }
      }
    }

    private var lastHighlightingAnchorIndex = 0
      set(value) {
        if (field != value) {
          field = value
          updateTexts()
        }
      }
    private val highlightingAnchorIndex: Int // highlight base on scrollY
      get() {
        val scrollY = container.scrollY
        anchors.forEachIndexed { index, anchor ->
          if (anchor.toScrollY >= scrollY) {
            return if (anchor.toScrollY - scrollY <= container.viewport.height / 2) {
              index
            } else {
              (index - 1).coerceAtLeast(0)
            }
          }
        }
        return anchors.size - 1
      }
    private val highlightingAnchor
      get() = if (anchors.isEmpty()) dummyAnchor else anchors[highlightingAnchorIndex]
    private val highlightingRowLastAnchor: Anchor
      get() = highlightingAnchorIndex.let { startIndex ->
        if (anchors.isEmpty()) return dummyAnchor
        for (i in startIndex + 1 until anchors.size) {
          if (anchors[i].rowIndex != anchors[startIndex].rowIndex) return anchors[i - 1]
        }
        return anchors.last()
      }

    override fun render(mouseX: Int, mouseY: Int, partialTicks: Float) {
      expanded = isMouseOver(mouseX, mouseY) && anchorsManager.totalTextRow > 1
      val (x1, y1, x2, y2) = absoluteBounds
      h(x1, x2 - 1, y1, if (expanded) COLOR_ANCHOR_BORDER_HOVER else COLOR_ANCHOR_BORDER)
      h(x1, x2 - 1, y2, if (expanded) COLOR_ANCHOR_BORDER_HOVER else COLOR_ANCHOR_BORDER)
      fill(x1, y1 + 1, x2, y2, if (expanded) COLOR_ANCHOR_BG_HOVER else COLOR_ANCHOR_BG)
      lastHighlightingAnchorIndex = highlightingAnchorIndex
      if (!expanded && anchorsManager.totalTextRow > 1) {
        highlightingRowLastAnchor.textButtonWidget.run {
          VanillaRender.drawString(" ... ...", absoluteBounds.right, screenY, COLOR_WHITE)
        }
      }
    }

    fun sizeChanged() { // re calc
      if (width != anchorsManager.width) {
        widgets.clear()
        anchorsManager.let { old ->
          anchorsManager = AnchorsManager().also { new ->
            old.anchors.forEach { new.addAnchor(it.anchorDisplayText, it.toScrollY) }
          }
        }
      }
    }
  }

  override fun render(mouseX: Int, mouseY: Int, partialTicks: Float) {
    if (renderBorder) {
      outline(absoluteBounds, borderColor)
    }
    super.render(mouseX, mouseY, partialTicks)
  }

  fun addEntry(entry: Widget) {
    containerContentFlowLayout.add(entry, entry.height)
    container.contentHeight = containerContentFlowLayout.offset
  }

  fun addAnchor(displayText: String) {
    anchorHeader.addAnchor(displayText)
  }

  fun isOutOfContainer(entry: Widget): Boolean =
      container.scrollY > entry.bounds.bottom || container.scrollY + container.viewport.height < entry.top

}